src/fsharp/TypeChecker.fs

```diff
@@ -565,13 +570,34 @@ type cenv =
       nameResolver: NameResolver
       
       conditionalDefines: string list
+
+      stringFormatEnv: Lazy<StringFormatEnv>
+
+      exprParser: string -> SynExpr
             
     } 
 
-    static member Create (g, isScript, niceNameGen, amap, topCcu, isSig, haveSig, conditionalDefines, tcSink, tcVal) =
+    static member Create (g, nenv, exprParser, isScript, niceNameGen, amap, topCcu, isSig, haveSig, conditionalDefines, tcSink, tcVal) =
         let infoReader = new InfoReader(g, amap)
         let instantiationGenerator m tpsorig = ConstraintSolver.FreshenTypars m tpsorig
         let nameResolver = new NameResolver(g, amap, infoReader, instantiationGenerator)
+        let stringFormatEnv = 
+            lazy
+                let concat =
+                    AllMethInfosOfTypeInScope infoReader nenv (Some("Concat"), AccessibleFromEverywhere) IgnoreOverrides range.Zero g.string_ty
+                    |> List.find (fun meth -> 
+                        match meth with
+                        | ILMeth(_, ilMethInfo, _) ->
+                            match ilMethInfo.ParamMetadata with
+                            | [{ Type = ILType.Array(shape, ty) }] -> shape = ILArrayShape.SingleDimensional && isILObjectTy ty
+                            | _ -> false
+                        | _ -> false
+                        )
+                let tostring =
```

Comment:
per @vasily-kirichenko's comment on the [RFC discussion thread](https://github.com/fsharp/fslang-design/issues/6#issuecomment-273456730), would it be possible here to detect if the type of the current expression is an F# type and use the equivalent of the `%A` format string instead of the `ToString` method for the type?

Reply:
If it produces better results I'll definitely do so. I know thanks to you already the details why it should be different treated. To implement it I would need to know from someone which is familiar with the compiler whether I can just check if a type is an F# type.

Reply:
Specifically only do this if ToString *hasn't* been overriden by the user.

Reply:
I don't like the current _casting code_ too. I would like to change it as soon as I found out how to inject format functions into the casting.